File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        string answer = "";
        int logest = 0;
        int logest_left = 0;
        int logest_right = 0;
        int cur_num = 0;
        int left;
        int right;

        for (auto i = 0; i < s.size(); i++){
            left = i - 1;
            right = i + 1;
            cur_num = 1;
            while(left >= 0 && right < s.size() && s[left] == s[right]){
                cur_num += 2;
                left--;
                right++;
            }

            if (cur_num > logest){
                logest = cur_num;
                logest_left = ++left;
                logest_right = --right;
            }
        }
        for (auto i = 0; i < s.size(); i++){
            left = i;
            right = i + 1;
            cur_num = 0;
            while(left >= 0 && right < s.size() && s[left] == s[right]){
                cur_num += 2;
                left--;
                right++;
            }

            if (cur_num > logest){
                logest = cur_num;
                logest_left = ++left;
                logest_right = --right;
            }
        }

        return s.substr(logest_left, logest_right-logest_left+1);
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        # brute force
        # start with a char extend to left and right and find the max possible palindromic string
        # start with a pair of char and extend left and right max possible string
        # time complexity o(2 * n * n) = o(n ^ 2)

        result = ""
        n = len(s)

        for i in range(len(s)):
            left = i
            right = i
            while left >= 0 and right < n:
                if s[left] == s[right]:
                    result = s[left: right + 1] if len(result) < right - left + 1 else result
                    left -= 1
                    right += 1
                else:
                    break
            
        for i in range(len(s) - 1):
            left = i
            right = i + 1

            while left >= 0 and right < n:
                if s[left] == s[right]:
                    result = s[left: right + 1] if len(result) < right - left + 1 else result
                    left -= 1
                    right += 1
                else:
                    break
        
        return result

        